File: sections/64_multiplicative_number_theory/problems/c_lcm_pair_spectrum/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using int64 = long long;
constexpr int64 MOD = 1'000'000'007;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, maximum;
    if (!(cin >> n >> maximum)) return 0;
    vector<int> frequency(maximum + 1);
    for (int i = 0, value; i < n; ++i) {
        cin >> value;
        ++frequency[value];
    }

    vector<int> inside(maximum + 1);
    for (int divisor = 1; divisor <= maximum; ++divisor)
        for (int multiple = divisor; multiple <= maximum; multiple += divisor)
            inside[multiple] += frequency[divisor];

    vector<int64> exact(maximum + 1);
    for (int value = 1; value <= maximum; ++value)
        exact[value] = 1LL * inside[value] * (inside[value] - 1) / 2 % MOD;
    for (int divisor = 1; divisor <= maximum; ++divisor)
        for (int multiple = 2 * divisor; multiple <= maximum; multiple += divisor) {
            exact[multiple] -= exact[divisor];
            if (exact[multiple] < 0) exact[multiple] += MOD;
        }

    for (int value = 1; value <= maximum; ++value)
        cout << exact[value] << (value == maximum ? '\n' : ' ');
}

File: sections/64_multiplicative_number_theory/problems/c_lcm_pair_spectrum/solution.py
import sys

MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, maximum = data[:2]
    frequency = [0] * (maximum + 1)
    for value in data[2 : 2 + n]:
        frequency[value] += 1

    inside = [0] * (maximum + 1)
    for divisor in range(1, maximum + 1):
        amount = frequency[divisor]
        if amount:
            for multiple in range(divisor, maximum + 1, divisor):
                inside[multiple] += amount

    exact = [count * (count - 1) // 2 % MOD for count in inside]
    for divisor in range(1, maximum + 1):
        contribution = exact[divisor]
        for multiple in range(2 * divisor, maximum + 1, divisor):
            exact[multiple] = (exact[multiple] - contribution) % MOD
    print(*exact[1:])


if __name__ == "__main__":
    main()
